File: src/displays/ssd1306.py
from smbus2 import SMBus
from PIL import Image, ImageDraw, ImageFont
import os

class SSD1306:
    """Driver for SSD1306 OLED display."""
    
    def __init__(self, bus):
        """Initialize the SSD1306 OLED display.
        
        Args:
            bus (int): I2C bus number (0 or 1)
        """
        self.bus = SMBus(bus)
        self.address = 0x3C  # Standard I2C address for SSD1306
        self.width = 128  # OLED display width
        self.height = 64  # OLED display height
        self.pages = self.height // 8
        
        # Create image buffer
        self.image = Image.new('1', (self.width, self.height))
        self.draw = ImageDraw.Draw(self.image)
        # Load a font with precise sizing for 15px height
        font_size = 14  # Starting with 14pt to achieve ~15px height
        try:
            self.font = ImageFont.truetype("/usr/share/fonts/truetype/liberation/LiberationSans-Regular.ttf", font_size)
        except:
            try:
                self.font = ImageFont.truetype("/usr/share/fonts/truetype/ubuntu/Ubuntu-R.ttf", font_size)
            except:
                try:
                    self.font = ImageFont.truetype("/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf", font_size)
                except:
                    self.font = ImageFont.load_default()
        
        # Calculate font dimensions using a test string
        test_text = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
        bbox = self.font.getbbox(test_text)
        self.font_height = bbox[3] - bbox[1]  # bottom - top
        self.font_width = (bbox[2] - bbox[0]) / len(test_text)  # average character width
        
        # Verify the height and adjust if needed
        while self.font_height != 15 and isinstance(self.font, ImageFont.FreeTypeFont):
            if self.font_height < 15:
                font_size += 1
            else:
                font_size -= 1
            self.font = ImageFont.truetype(self.font.path, font_size)
            bbox = self.font.getbbox(test_text)
            self.font_height = bbox[3] - bbox[1]
        
        # Calculate line spacing
        self.line_spacing = self.font_height + 1  # Add 1 pixel vertical interval
        
        self.initialize_display()
    
    def initialize_display(self):
        """Initialize the display with required settings."""
        commands = [
            0xAE,        # Display off
            0xD5, 0x80,  # Set display clock div
            0xA8, 0x1F,  # Set multiplex to 32 (0x1F + 1)
            0xD3, 0x00,  # Set display offset
            0x40,        # Set start line to 0
            0x8D, 0x14,  # Charge pump setting
            0x20, 0x00,  # Memory mode
            0xA1,        # Segment remap
            0xC8,        # COM scan direction
            0xDA, 0x02,  # Set COM pins hardware configuration
            0x81, 0xCF,  # Set contrast
            0xD9, 0xF1,  # Set precharge
            0xDB, 0x40,  # Set vcom detect
            0xA4,        # Display all on resume
            0xA6,        # Normal display
            0x2E,        # Deactivate scroll
            0xAF         # Display on
        ]
        
        # Send commands properly
        for cmd in commands:
            self._write_command(cmd)
            
    def _write_command(self, command):
        """Write a single command to the display."""
        try:
            self.bus.write_byte_data(self.address, 0x00, command)
        except Exception as e:
            print(f"Error writing command {hex(command)}: {e}")
            
    def _write_data(self, data):
        """Write data to the display."""
        try:
            self.bus.write_byte_data(self.address, 0x40, data)
        except Exception as e:
            print(f"Error writing data: {e}")

    def clear_display(self):
        """Clear the display."""
        self.image.paste(0, (0, 0, self.width, self.height))
        self._display_image()
    
    def display_measurements(self, pressure, altitude, temperature, acceleration):
        """Display the four measurements on the OLED screen in two rows.
        Each row is 15 pixels high with 1 pixel interval between rows.
        First row: 0-14 pixels
        Interval: 16th pixel
        Second row: 17-31 pixels
        
        Args:
            pressure (float): Pressure value in hPa
            altitude (float): Altitude in meters
            temperature (float): Temperature in Celsius
            acceleration (float): Acceleration in m/s²
        """
        # Clear the image
        self.draw.rectangle((0, 0, self.width, self.height), outline=0, fill=0)
        
        # Format the values with proper units (shortened for space)
        p_text = f"P:{pressure:.2f}"
        ac_text = f"AC:{acceleration:.1f}"
        t_text = f"T:{temperature:.1f}"
        a_text = f"A:{altitude:.0f}m"
        
        # Position text with 15px height rows and 1px interval
        y1 = 0    # First row starts at 0
        y2 = 17   # Second row starts at 17 (after 15px height + 1px interval)
        
        self.draw.text((5, y1), p_text, font=self.font, fill=255)
        self.draw.text((self.width//2 + 10, y1), ac_text, font=self.font, fill=255)
        self.draw.text((5, y2), t_text, font=self.font, fill=255)
        self.draw.text((self.width//2 - 5, y2), a_text, font=self.font, fill=255)
        
        self._display_image()
    
    def _display_image(self):
        """Internal method to send the image buffer to the display."""
        # Set column address
        self._write_command(0x21)  # Column address command
        self._write_command(0)     # Start column
        self._write_command(127)   # End column
        
        # Set page address
        self._write_command(0x22)  # Page address command
        self._write_command(0)     # Start page
        self._write_command(7)     # End page
        
        # Convert image to display format
        pixels = self.image.convert("1")
        buffer = bytearray(self.width * self.pages)
        
        # Arrange pixels into pages
        for y in range(self.height):
            for x in range(self.width):
                if pixels.getpixel((x, y)):
                    buffer[x + (y // 8) * self.width] |= (1 << (y % 8))
        
        # Write buffer to display in chunks
        chunk_size = 32  # Write 32 bytes at a time
        for i in range(0, len(buffer), chunk_size):
            chunk = buffer[i:i + chunk_size]
            self.bus.write_i2c_block_data(self.address, 0x40, list(chunk)) 

File: src/drivers/ssd1306/ssd1306.c
#include "drivers/ssd1306/ssd1306.h"
#include "drivers/bus/i2c_bus.h" // Include necessary headers
#include "common/common_types.h"        // Include necessary headers
#include <stdlib.h> // For malloc, free
#include <string.h> // For memset, memcpy
#include <stdio.h>  // For fprintf, stderr
#include <unistd.h> // For usleep

// --- SSD1306 Command Definitions ---
// Fundamental Commands
#define SSD1306_SETCONTRAST 0x81
#define SSD1306_DISPLAYALLON_RESUME 0xA4 // Resume normal display output from RAM
#define SSD1306_DISPLAYALLON 0xA5        // Force entire display ON, ignoring RAM
#define SSD1306_NORMALDISPLAY 0xA6
#define SSD1306_INVERTDISPLAY 0xA7
#define SSD1306_DISPLAYOFF 0xAE
#define SSD1306_DISPLAYON 0xAF

// Addressing Setting Commands
#define SSD1306_MEMORYMODE 0x20
#define SSD1306_COLUMNADDR 0x21 // Requires 2 more bytes (start, end)
#define SSD1306_PAGEADDR 0x22   // Requires 2 more bytes (start, end)
// For MEMORYMODE 0x00 (Horizontal): Col Address Pointer increases, wraps page, Page Address Pointer resets
// For MEMORYMODE 0x01 (Vertical): Page Address Pointer increases, wraps column, Col Address Pointer resets
// For MEMORYMODE 0x02 (Page): Col Address Pointer increases, remains on same page (DEFAULT)

// Hardware Configuration Commands
#define SSD1306_SETSTARTLINE 0x40 // OR with 0-63
#define SSD1306_SEGREMAP 0xA0     // OR with 0x01 for column 127 mapped to SEG0
#define SSD1306_SETMULTIPLEX 0xA8 // Requires 1 more byte (15-63)
#define SSD1306_COMSCANINC 0xC0   // Normal COM0...COM[N-1] scan
#define SSD1306_COMSCANDEC 0xC8   // Remapped COM[N-1]...COM0 scan
#define SSD1306_SETDISPLAYOFFSET 0xD3 // Requires 1 more byte (0-63)
#define SSD1306_SETCOMPINS 0xDA       // Requires 1 more byte

// Timing & Driving Scheme Commands
#define SSD1306_SETDISPLAYCLOCKDIV 0xD5 // Requires 1 more byte
#define SSD1306_SETPRECHARGE 0xD9       // Requires 1 more byte
#define SSD1306_SETVCOMDETECT 0xDB      // Requires 1 more byte

// Charge Pump Command
#define SSD1306_CHARGEPUMP 0x8D // Requires 1 more byte (0x10 disable, 0x14 enable)

// Define a chunk size for I2C transfers
#define SSD1306_I2C_CHUNK_SIZE 64 // Send 64 bytes of pixel data at a time

// Command to set Page Start Address for Page Addressing Mode
#define SSD1306_SETPAGESTARTADDR 0xB0 // OR with page number 0-7

// --- Internal Helper Functions ---

// --- Static Helper Function Forward Declarations ---
// Add these prototypes before any function implementations
static status_t ssd1306_send_command(ssd1306_handle_t* handle, uint8_t command);
static status_t ssd1306_send_data(ssd1306_handle_t* handle, const uint8_t* data, size_t size);

/**
 * @brief Sends a single command byte to the SSD1306.
 * @param handle Pointer to the SSD1306 handle.
 * @param command The command byte to send.
 * @return STATUS_OK on success, or an I2C error code.
 */
static status_t ssd1306_send_command(ssd1306_handle_t* handle, uint8_t command) {
    if (!handle || !handle->i2c_handle) {
        return ERROR_NULL_POINTER;
    }
    // For SSD1306 over I2C, commands are sent as data bytes following
    // a control byte with the Co (Continuation) bit 0 and D/C# (Data/Command) bit 0.
    // Control byte = 0x00
    uint8_t buffer[2] = {0x00, command};
    // Retry I2C write up to 3 times with small delay to avoid timeouts
    status_t status;
    for (int retry = 0; retry < 3; ++retry) {
        status = i2c_bus_write(handle->i2c_handle, handle->i2c_addr, buffer, 2);
        if (status == STATUS_OK) break;
        usleep(100);
    }
    return status;
}

// --- Public Function Implementations ---

status_t ssd1306_init(i2c_bus_handle_t i2c_bus, uint8_t i2c_address, uint8_t width, uint8_t height, ssd1306_handle_t** out_handle) {
    if (!i2c_bus || !out_handle) {
        return ERROR_NULL_POINTER;
    }
    if (width == 0 || height == 0 || (width * height / 8) == 0) {
        return ERROR_INVALID_PARAM;
    }

    status_t status = STATUS_OK;
    ssd1306_handle_t* handle = NULL;

    // 1. Allocate memory for the handle
    handle = (ssd1306_handle_t*)malloc(sizeof(ssd1306_handle_t));
    if (!handle) {
        return ERROR_NO_MEM;
    }
    memset(handle, 0, sizeof(ssd1306_handle_t)); // Clear handle memory

    // 2. Store configuration
    handle->i2c_handle = i2c_bus;
    handle->i2c_addr = i2c_address;
    handle->width = width;
    handle->height = height;
    handle->buffer_size = (size_t)width * height / 8;

    // 3. Allocate memory for the display buffer
    handle->buffer = (uint8_t*)malloc(handle->buffer_size);
    if (!handle->buffer) {
        fprintf(stderr, "ERROR: ssd1306_init: Failed to allocate display buffer (%zu bytes)\n", handle->buffer_size);
        free(handle); // Clean up handle allocation
        return ERROR_NO_MEM;
    }
    memset(handle->buffer, 0, handle->buffer_size); // Clear buffer initially

    // 4. Send COMPREHENSIVE initialization sequence
    printf("INFO: ssd1306_init: Sending initialization sequence ...\n");

    // Initialization Sequence using Adafruit's standard configuration
    const uint8_t init_sequence[] = {
        SSD1306_DISPLAYOFF,         // 0xAE - Display OFF

        SSD1306_SETDISPLAYCLOCKDIV, // 0xD5 - Set Clock Divide Ratio/Oscillator Frequency
        0x80,                       //   Default ratio

        SSD1306_SETMULTIPLEX,       // 0xA8 - Set Multiplex Ratio
        (uint8_t)(height - 1),      //   height - 1 (e.g., 63 for 128x64)

        SSD1306_SETDISPLAYOFFSET,   // 0xD3 - Set Display Offset
        0x00,                       //   No offset

        SSD1306_SETSTARTLINE | 0x0, // 0x40 - Set Display Start Line (line 0)

        SSD1306_CHARGEPUMP,         // 0x8D - Charge Pump Setting
        0x14,                       //   Enable charge pump (for internal VCC)

        SSD1306_MEMORYMODE,         // 0x20 - Set Memory Addressing Mode
        0x00,                       //   <<< 0x00 = Horizontal Addressing Mode >>>

        SSD1306_SEGREMAP | 0x1,     // 0xA1 - Set Segment Re-map (col 127 mapped to SEG0)
                                    //   This flips the display horizontally

        SSD1306_COMSCANDEC,         // 0xC8 - Set COM Output Scan Direction (scan COM[N-1] to COM0)
                                    //   This flips the display vertically

        SSD1306_SETCOMPINS,         // 0xDA - Set COM Pins Hardware Configuration
        (height == 64) ? 0x12 : 0x02, // 0x12 for 128x64, 0x02 for 128x32 (common values)

        SSD1306_SETCONTRAST,        // 0x81 - Set Contrast Control
        0xCF,                       //   Contrast value (adjust as needed)

        SSD1306_SETPRECHARGE,       // 0xD9 - Set Pre-charge Period
        0xF1,                       //   Pre-charge period (for internal VCC)

        SSD1306_SETVCOMDETECT,      // 0xDB - Set VCOMH Deselect Level
        0x40,                       //   VCOM deselect level (adjust as needed)

        SSD1306_DISPLAYALLON_RESUME,// 0xA4 - Entire Display ON (resume from RAM content)
        SSD1306_NORMALDISPLAY,      // 0xA6 - Set Normal Display mode

        SSD1306_DISPLAYON           // 0xAF - Display ON in normal mode
    };

    for (size_t i = 0; i < sizeof(init_sequence); ++i) {
        status = ssd1306_send_command(handle, init_sequence[i]);
        if (status != STATUS_OK) {
            fprintf(stderr, "ERROR: ssd1306_init: Failed command 0x%02X (index %zu), status=%d\n",
                    init_sequence[i], i, status);
            // Go to cleanup if allocation succeeded before failure
            goto error_cleanup;
        }
        usleep(100); // Small delay between commands might help stability
    }
    usleep(100 * 1000); // Delay after sequence

    // Set full-screen addressing window (horizontal mode)
    ssd1306_send_command(handle, SSD1306_PAGEADDR);
    ssd1306_send_command(handle, 0);
    ssd1306_send_command(handle, (handle->height / 8) - 1);
    ssd1306_send_command(handle, SSD1306_COLUMNADDR);
    ssd1306_send_command(handle, 0);
    ssd1306_send_command(handle, handle->width - 1);

    // Clear buffer and display initially
    status = ssd1306_clear_buffer(handle);
    if (status != STATUS_OK) { /* Handle error or warning */ }
    status = ssd1306_display(handle); // Send empty buffer
    if (status != STATUS_OK) { /* Handle error or warning */ }
    *out_handle = handle;
    printf("INFO: ssd1306_init: Initialization successful.\n");
    return STATUS_OK;

error_cleanup: // Label for cleaning up resources on error during init
    fprintf(stderr, "ERROR: ssd1306_init: Cleaning up after initialization failure.\n");
    if (handle) {
        if (handle->buffer) {
            free(handle->buffer);
            handle->buffer = NULL;
        }
        free(handle);
    }
    *out_handle = NULL;
    return status; // Return the error status that caused the jump
}

// --- Other Public Functions (To be implemented) ---

void ssd1306_cleanup(ssd1306_handle_t* handle) {
    if (!handle) {
        return; // Nothing to clean up if handle is NULL
    }

    // Optional: Turn the display off before cleaning up
    // Ignore potential errors here, as we are cleaning up anyway
    (void)ssd1306_send_command(handle, SSD1306_DISPLAYOFF);

    // Free the display buffer if it was allocated
    if (handle->buffer) {
        free(handle->buffer);
        handle->buffer = NULL; // Avoid dangling pointer
    }

    // Free the handle itself
    free(handle);

    printf("INFO: ssd1306_cleanup: Resources released.\n");
}

status_t ssd1306_clear_buffer(ssd1306_handle_t* handle) {
    if (!handle || !handle->buffer) return ERROR_NULL_POINTER;
    memset(handle->buffer, 0x00, handle->buffer_size); // Set all pixels OFF
    return STATUS_OK;
}

status_t ssd1306_draw_pixel(ssd1306_handle_t* handle, int16_t x, int16_t y, uint8_t color) {
    if (!handle || !handle->buffer) {
        return ERROR_NULL_POINTER;
    }
    // Check bounds
    if (x < 0 || x >= handle->width || y < 0 || y >= handle->height) {
        return ERROR_INVALID_PARAM; // Out of bounds
    }

    // Calculate buffer index based on SSD1306 page addressing:
    // Index = column + (page_number * columns_per_page)
    uint16_t buffer_index = x + (y / 8) * handle->width;

    // Calculate the specific bit within the byte to modify.
    uint8_t bit_mask = 1 << (y % 8);

    // Modify the buffer
    if (color) { // Set pixel
        handle->buffer[buffer_index] |= bit_mask;
    } else { // Clear pixel
        handle->buffer[buffer_index] &= ~bit_mask;
    }
    return STATUS_OK;
}

status_t ssd1306_display(ssd1306_handle_t* handle) {
    if (!handle || !handle->buffer || !handle->i2c_handle) {
        return ERROR_NULL_POINTER;
    }
    // Only send raw frame data; window set during init
    status_t status = ssd1306_send_data(handle, handle->buffer, handle->buffer_size);
    if (status != STATUS_OK) {
        fprintf(stderr, "ERROR: ssd1306_display: Failed to write buffer data, status=%d\n", status);
    }
    return status;
}

status_t ssd1306_update(ssd1306_handle_t* handle) {
    // Send entire buffer to display
    return ssd1306_display(handle);
}

// --- Helper Functions ---

// Sends a buffer of data bytes, chunking if necessary
static status_t ssd1306_send_data(ssd1306_handle_t* handle, const uint8_t* data, size_t size) {
    if (!handle || !handle->i2c_handle || !data) return ERROR_NULL_POINTER;
    if (size == 0) return STATUS_OK;

    // Full-buffer write for maximum throughput (~40 fps)
    uint8_t* full_buf = malloc(size + 1);
    if (!full_buf) return ERROR_NO_MEM;
    full_buf[0] = 0x40; // Data control byte
    memcpy(full_buf + 1, data, size);
    status_t status = i2c_bus_write(handle->i2c_handle, handle->i2c_addr,
                                    full_buf, size + 1);
    free(full_buf);
    return status;
}